### PR TITLE
Add support for enums in @prop decorator

### LINKED ISSUE #31: Enhanced enum support
Right now, you support string array as an enum:
```ts
// Enum-like type and definition example.
type Gender = 'male' | 'female';
const Genders = {
  MALE: 'male' as Gender,
  FEMALE: 'female' as Gender,
};

@prop({ enum: Object.values(Genders) })
gender?: Gender;
```
However we have string enum support in TS now. It would be nice if the decorator function accept also the `object` type as the `enum` value and get the string values using `Object.values(object)` inside:
```ts
// true enum type
enum Gender {
  MALE = 'male',
  FEMALE = 'female',
};

@prop({ enum: Gender })
gender?: Gender;
```
So the signature of the `BasePropOptions` would be `enum?: string[]|object;` and you can in runtime check if `Array.isArray(enum)` to behave like now or convert enum object to values.

### PR DESCRIPTION
**Pull request by 19majkel94**
Closes #31.

### CODE REVIEW

**Comment by 19majkel94** [comment]
I see that the build failed but I don't know how to configure local environment to run the tests. Could you check what caused the error in this simple PR? You can push fix into my branch as I checked "Allow edits from maintainers".

File: src/prop.ts [hunk 1]
```diff
@@ -11,7 +11,7 @@ export type RequiredType = boolean | [boolean, string] | string | Func | [Func,
 
 export interface BasePropOptions {
   required?: RequiredType;
-  enum?: string[];
+  enum?: string[]|object;
```
Comment (szokodiakos) [inline src/prop.ts]: apply spacing `enum?: string[] | object;`

File: src/test/index.test.ts [hunk 2]
```diff
@@ -8,6 +8,7 @@ import { expect } from 'chai';
 import { model as User } from './models/user';
 import { model as Car, Car as CarType } from './models/car';
 import { Genders } from './enums/genders';
+import { Role } from "./enums/role";
```
Comment (szokodiakos) [inline src/test/index.test.ts]: use singlequote please

File: src/test/enums/role.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export enum Role {
+    Admin = "admin",
```
Comment (szokodiakos) [inline src/test/enums/role.ts]: use singlequote in this file please

File: src/test/models/user.ts [hunk 4]
```diff
@@ -5,6 +5,7 @@ import * as findOrCreate from 'mongoose-findorcreate';
 import { Job } from './job';
 import { Car } from './car';
 import { Gender, Genders } from '../enums/genders';
+import { Role } from "../enums/role";
```
Comment (szokodiakos) [inline src/test/models/user.ts]: use singlequote please

**Comment by 19majkel94** [comment]
@szokodiakos Lint issues has been fixed.

**Comment by szokodiakos** [comment]
Thanks for the PR update! I will look into the build issue when I'll have the time.

**Comment by vinczedani** [comment]
Hi @19majkel94 
I checked your PR and to me it looks fine. The tests fail because you marked the role property as required. In the test there is more the one User created, but you only added a role to one of them.
Possible fixes can be:
- Removing the required flag from the prop.
- Setting a default value for the role prop.
- Adding roles to other users too.
Thanks for the contribution,
Daniel

**Comment by 19majkel94** [comment]
> Removing the required flag from the prop

That was the easiest one 😉 

@szokodiakos 
I also removed `libs` folder from repo + added it into `.gitignore` - we shouldn't hold transpiled files on repository for each commit.

**Comment by vinczedani** [comment]
You are right. We should improve the test coverage and quality by a lot. This should be one of our main priorities right now.
I alse like removing the lib folder, I hope @szokodiakos will approve the PR too, I do :+1:

**Comment by 19majkel94** [comment]
@vinczedani 
Done 😉

**Comment by thamera** [comment]
did this break?  I can't get an enum to work in @prop no mater what approach if I import from another file.  If I define the enum directly in the same file as my typegoose model it works.  Put the exact same enum as an export within it's own ts file and I get ann error "Type of role property is not a string."  I tried setting up that export both as a traditional enum and a string array.  I am on the latest typescript and typegoose versions.

**Comment by hasezoey** [comment]
@thamera if you have an issue, please open a new issue, and not write it on such an old issue/pr, thanks

**Pull request merged**